
configfile: "config/config.yaml"

# get paired end samples name
SAMPLES = config["samples"]

# get reference genome
genome = config["reference_genome"]


rule all:
    input:
        expand("results/{sample}/mapped/{sample}.sorted.bam.flagstat", sample=SAMPLES),
        expand("results/{sample}/mapped/{sample}.sorted.bam.idxstats", sample=SAMPLES),
        expand("results/{sample}/lineage_assignment/{sample}.csv", sample=SAMPLES),
        expand("results/{sample}/megahit_assembled/{sample}", sample=SAMPLES),
        expand("workflow/report/{sample}/{sample}_variants.html", sample=SAMPLES),
        expand("results/{sample}/vcf/{sample}.variants.annotated.snpsift.vcf.gz", sample=SAMPLES),
        #expand("results/{sample}/vcf/{sample}.dbnsfp.snpsift.vcf.gz", sample=SAMPLES),
        expand("results/{sample}/vcf/{sample}.varType.snpsift.vcf.gz.vcf", sample=SAMPLES),
        #expand("results/{sample}/vcf/{sample}.genesets.snpsift.vcf", sample=SAMPLES),
        
        #vep result
       # expand("results/{sample}/vcf/{sample}.variants.html", sample=SAMPLES)
        


include: "rules/preprocess.smk"
include: "rules/aligning.smk"
include: "rules/calling.smk"
include: "rules/lineage_assignment.smk"
include: "rules/denovo.smk"
include: "rules/generate_report.smk"
#include: "rules/vcf_variant_annotation.smk"
include: "rules/vcf_variant_annotation_snpsift.smk"